staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/vault/vault_test.go

```diff
@@ -0,0 +1,338 @@
+}
+
+func onceUseTokenServer(t *testing.T, forbidden *bool, encryptCount, decryptCount *int) *httptest.Server {
+	roleLogin := approleLoginHandler{t}
+	appRoleLoginFunc := func(w http.ResponseWriter, r *http.Request) {
+		*forbidden = false
+		roleLogin.ServeHTTP(w, r)
+	}
+
+	encrypt := encryptHandler{t}
+	encryptFunc := func(w http.ResponseWriter, r *http.Request) {
+		*encryptCount++
+		if !*forbidden {
+			encrypt.ServeHTTP(w, r)
+			*forbidden = true
+		} else {
+			http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+		}
+	}
+
+	decrypt := decryptHandler{t}
+	decryptFunc := func(w http.ResponseWriter, r *http.Request) {
+		*decryptCount++
+		if !*forbidden {
+			decrypt.ServeHTTP(w, r)
+			*forbidden = true
+		} else {
+			http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+		}
+	}
+
+	handlers := map[string]http.Handler{
+		"/v1/auth/approle/login": http.HandlerFunc(appRoleLoginFunc),
+		"/v1/transit/encrypt/":   http.HandlerFunc(encryptFunc),
+		"/v1/transit/decrypt/":   http.HandlerFunc(decryptFunc),
+	}
+
+	return VaultTestServer(t, handlers)
+}
+
+func TestInvalidConfiguration(t *testing.T) {
+	server := VaultTestServer(t, nil)
+	defer server.Close()
+
+	// No key name
+	configWithoutKey := `
+addr: @url@
+ca-cert: testdata/ca.crt
+token: 8dad1053-4a4e-f359-2eab-d57968eb277f
+`
+	// No address
+	configWithoutAddress := `
+key-names: 
+  - @key@
+ca-cert: testdata/ca.crt
+token: 8dad1053-4a4e-f359-2eab-d57968eb277f
+`
+	// No any authentication info
+	configWithoutAuth := `
+key-names: 
+  - @key@
+addr: @url@
+ca-cert: testdata/ca.crt
+`
+	// tls authentication, but no client key
+	configTLSWithoutClientKey := `
+key-names: 
+  - @key@
+addr: @url@
+ca-cert: testdata/ca.crt
+client-cert: testdata/client.crt
+`
+	// tls authentication, but no client cert
+	configTLSWithoutClientCert := `
+key-names: 
+  - @key@
+addr: @url@
+ca-cert: testdata/ca.crt
+client-key: testdata/client.key
+`
+	// approle authentication, but no role id
+	configRoleWithoutRoleID := `
+key-names: 
+  - @key@
+addr: @url@
+ca-cert: testdata/ca.crt
+secret-id: cd834818-ac2b-4db0-b3e9-2cdd6db599f0
+`
+	// there are more than one authentication
+	configMoreThanOneAuth := `
+key-names: 
+  - @key@
+addr: @url@
+ca-cert: testdata/ca.crt
+token: 8dad1053-4a4e-f359-2eab-d57968eb277f
+role-id: 655a9287-f1be-4be0-844c-4f13a1757532
+`
+
+	invalidConfigs := []struct {
+		config      string
+		description string
+	}{
+		{configWithoutKey, "there is no key name"},
+		{configWithoutAddress, "there is no address"},
+		{configWithoutAuth, "there is no authentication"},
+		{configTLSWithoutClientKey, "there is no client key file"},
+		{configTLSWithoutClientCert, "there is no client cert file"},
+		{configRoleWithoutRoleID, "there is no role id"},
+		{configMoreThanOneAuth, "there are more than one authentications"},
+	}
+
+	for _, testCase := range invalidConfigs {
+		_, err := serviceTestFactory(testCase.config, server.URL, key)
```

Comment:
Should use https://golang.org/pkg/testing/#T.Run

Reply:
fixed